src/dtx/dtx_common.c

```diff
@@ -405,6 +405,14 @@ dtx_end(struct dtx_handle *dth, struct ds_cont_hdl *cont_hdl,
 
 	if (dth->dth_leader && dth->dth_dti_cos != NULL)
 		D_FREE(dth->dth_dti_cos);
+	while (d_list_empty(&dth->dth_shares)) {
```

Comment:
I think this would be cleaner if you use d_list_pop_entry instead

while ((dts = d_list_pop_entry(dth->dth_shares, struct dtx_share, dts_link) != NULL)
      D_FREE_PTR(dts);